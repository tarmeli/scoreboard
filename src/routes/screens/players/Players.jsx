import React from "react";
import { Button, Div, Form, Input, Table } from "../../../components/Element/";
import { Tbody, Td, Tr } from "../../../components/Element/Table/";
import Map from "../../../components/utils/Map/Map";
import PlayersProvider from "./context/provider/PlayersProvider";
import PlayersHook from "./context/hook/PlayersHook";
import PlayersTableHeader from "./PlayersTableHeader/PlayersTableHeader";
import LoadingTentative from "../../../components/LoadingTentative/LoadingTentative";

const Players = ({ data: { players } }) => {
  return (
    <PlayersProvider>
      <PlayersHook>
        {({
          isLoading,
          handleDecrementWins,
          handleIncrementWins,
          handleDecrementLosses,
          handleIncrementLosses,
          handleAddNewPlayer,
          handleRemovePlayer,
          newPlayerNameInputValue,
          handleNewPlayerNameChange,
          submitNewPlayerButtonIsDisabled,
        }) => (
          <Div flexContainer flexColumn>
            <Table flexItem>
              <PlayersTableHeader />

              <Tbody>
                <Map items={players}>
                  {({ name, wins, losses, _id: id }) => {
                    return (
                      <Tr>
                        <Td>{name}</Td>

                        <Td>
                          <Button
                            onClick={async () => await handleDecrementWins(id)}
                          >
                            -
                          </Button>
                        </Td>

                        <Td>{wins}</Td>

                        <Td>
                          <Button
                            onClick={async () => await handleIncrementWins(id)}
                          >
                            +
                          </Button>
                        </Td>

                        <Td />

                        <Td>
                          <Button
                            onClick={async () =>
                              await handleDecrementLosses(id)
                            }
                          >
                            -
                          </Button>
                        </Td>

                        <Td>{losses}</Td>

                        <Td>
                          <Button
                            onClick={async () =>
                              await handleIncrementLosses(id)
                            }
                          >
                            +
                          </Button>
                        </Td>

                        <Td>
                          <Button
                            fullWidth
                            onClick={async () => await handleRemovePlayer(id)}
                          >
                            <LoadingTentative isLoading={isLoading}>
                              Remove Player
                            </LoadingTentative>
                          </Button>
                        </Td>
                      </Tr>
                    );
                  }}
                </Map>
              </Tbody>
            </Table>

            <Form
              flexItem
              fullWidth
              onSubmit={async () => await handleAddNewPlayer()}
            >
              <Div flexContainer>
                <Input
                  flexItem
                  fullWidth
                  value={newPlayerNameInputValue}
                  onChange={handleNewPlayerNameChange}
                />

                <Button
                  flexItem
                  fullWidth
                  type="submit"
                  disabled={submitNewPlayerButtonIsDisabled || isLoading}
                >
                  <LoadingTentative isLoading={isLoading}>
                    Add Player
                  </LoadingTentative>
                </Button>
              </Div>
            </Form>
          </Div>
        )}
      </PlayersHook>
    </PlayersProvider>
  );
};

export default Players;
