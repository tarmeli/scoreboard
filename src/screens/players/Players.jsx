import React from "react";
import Div from "../../components/Element/Div/Div";

import players from "../../data/stubs/players";
import Ul from "../../components/Element/List/Ul/Ul";
import Map from "../../components/Utility/Map/Map";
import Li from "../../components/Element/List/Li/Li";

const Players = () => (
    <Div>
        <Ul>
            <Map items={players}>
                {({ name, wins, losses }) => (
                    <Li>
                        {name}, wins {wins} - losses {losses}
                    </Li>
                )}
            </Map>
        </Ul>
    </Div>
);

export default Players;
