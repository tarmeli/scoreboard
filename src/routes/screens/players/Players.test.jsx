import React, { useState as useStateMock } from "react";
import { mount } from "enzyme";

import Players from "./Players";

describe("Players", () => {
  let component;
  let setStateMock;

  beforeEach(() => {
    setStateMock = jest.fn();
    useStateMock.mockImplementation((init) => [init, setStateMock]);

    component = mount(
      <Players
        data={{
          players: [
            {
              name: "Some player",
              wins: 42,
              losses: 84,
              _id: "some-player-id",
            },
            {
              name: "Some other player",
              wins: 126,
              losses: 168,
              _id: "some-other-player-id",
            },
          ],
        }}
      />
    );
  });

  it("renders", () => {
    expect(component.render()).toMatchSnapshot();
  });
});
