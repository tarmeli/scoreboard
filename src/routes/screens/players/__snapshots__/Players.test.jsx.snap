// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Players renders 1`] = `
<div>
  <div>
    <div>
      <table>
        <thead>
          <tr>
            <th>
              Player name
            </th>
            <th />
            <th>
              Wins
            </th>
            <th />
            <th />
            <th />
            <th>
              Losses
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Some player
            </td>
            <td>
              <button
                class="button"
              >
                -
              </button>
            </td>
            <td>
              42
            </td>
            <td>
              <button
                class="button"
              >
                +
              </button>
            </td>
            <td />
            <td>
              <button
                class="button"
              >
                -
              </button>
            </td>
            <td>
              84
            </td>
            <td>
              <button
                class="button"
              >
                +
              </button>
            </td>
            <td>
              <button
                class="button"
              >
                Remove Player
              </button>
            </td>
          </tr>
          <tr>
            <td>
              Some other player
            </td>
            <td>
              <button
                class="button"
              >
                -
              </button>
            </td>
            <td>
              126
            </td>
            <td>
              <button
                class="button"
              >
                +
              </button>
            </td>
            <td />
            <td>
              <button
                class="button"
              >
                -
              </button>
            </td>
            <td>
              168
            </td>
            <td>
              <button
                class="button"
              >
                +
              </button>
            </td>
            <td>
              <button
                class="button"
              >
                Remove Player
              </button>
            </td>
          </tr>
        </tbody>
      </table>
      <form>
        <input
          value=""
        />
        <button
          class="button"
          disabled=""
          type="submit"
        >
          Add Player
        </button>
      </form>
    </div>
  </div>
</div>
`;
